src/Umbraco.Tests/FrontEnd/UmbracoHelperTests.cs

```diff
@@ -78,5 +93,76 @@ public void Create_Encrypted_RouteString_From_Dictionary()
             
             Assert.AreEqual(expectedResult, result);
         }
+
+        [Test]
+        public void Truncate_By_Words_TagsAsContentOff()
+        {
+            var text = "Hello world, this is some text <a href='blah'>with a link</a>";
+
+            var helper = new UmbracoHelper();
+
+            var result = helper.TruncateByWords(text, 4, true, false).ToString();
+
+            Assert.AreEqual("Hello world, this is&hellip;", result);
+        }
+
+        [Test]
+        public void Truncate_By_Words_With_Tag_TagsAsContentOff()
+        {
+            var text = "Hello world, <b>this</b> is some text <a href='blah'>with a link</a>";
+
+            var helper = new UmbracoHelper();
+
+            var result = helper.TruncateByWords(text, 4, true, false).ToString();
+
+            Assert.AreEqual("Hello world, <b>this</b> is&hellip;", result);
+        }
+
+        [Test]
+        public void Truncate_By_Words_Mid_Tag_TagsAsContentOff()
+        {
+            var text = "Hello world, this is some text <a href='blah'>with a link</a>";
+
+            var helper = new UmbracoHelper();
+
+            var result = helper.TruncateByWords(text, 7, true, false).ToString();
+
+            Assert.AreEqual("Hello world, this is some text <a href='blah'>with&hellip;</a>", result);
+        }
+        [Test]
+        public void Truncate_By_Words_TagsAsContentOn()
+        {
+            var text = "Hello world, this is some text <a href='blah'>with a link</a>";
+
+            var helper = new UmbracoHelper();
+
+            var result = helper.TruncateByWords(text, 4, true, true).ToString();
+
+            Assert.AreEqual("Hello world, this is&hellip;", result);
+        }
+
+        [Test]
+        public void Truncate_By_Words_With_Tag_TagsAsContentOn()
+        {
+            var text = "Hello world, <b>this</b> is some text <a href='blah'>with a link</a>";
+
+            var helper = new UmbracoHelper();
+
+            var result = helper.TruncateByWords(text, 4, true, true).ToString();
+
+            Assert.AreEqual("Hello world, <b>this</b> is&hellip;", result);
+        }
+
+        [Test]
+        public void Truncate_By_Words_Mid_Tag_TagsAsContentOn()
```

Comment:
There seems to be tests missing for testing invalid html parsing and the logic that is in place to work around those types of problems.